Add examples to exportparts

KonnorRogers opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

Adding examples for using exportparts

### Motivation

The exportparts documentation has no examples and simply links to the spec and is not very welcoming to new users.

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

KonnorRogers committed:
Add examples to exportparts

KonnorRogers committed:
Update index.md

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -17,6 +17,76 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+class CardComponent extends HTMLElement {
+  connectedCallback() {
+    this.attachShadow({ mode: "open" });
+    this.shadowRoot.innerHTML = `
```
Let's use HTML to create HTML. This is a perfect opportunity to demonstrate using `<template>`

KonnorRogers replied:
I'm not sure I follow.

Something like this?

```html
<template id="card-component-template">
  <style>
    .base { display: grid };
  </style>
       
   <div class="base" part="base">
     <div part="header"><slot name="header"></slot></div>
     <div part="body"><slot></slot></div>
     <div part="footer"><slot name="footer"></slot></div>
    </div>
</template>
```

```js
class CardComponent extends HTMLElement {
  connectedCallback () {
    this.attachShadow({ mode: "open" })
    const firstChild = this.shadowRoot.firstElementChild
    const content = document.querySelector("#card-component-template").content.cloneNode(true)
    
   this.shadowRoot.innerHTML = ""
    
    this.shadowRoot.appendChild(content)
  }
}
```

estelle replied:
Yes. Exactly. Especially the HTML part.

For the HTML pages, where this example resides, I thought it's preferable to use the [`define`](https://developer.mozilla.org/en-US/docs/Web/API/Window/customElements/define) method of the Window's [customElements](https://developer.mozilla.org/en-US/docs/Web/API/Window/customElements) as it's [defined in the WHATWG spec](https://html.spec.whatwg.org/multipage/custom-elements.html#dom-window-customelements). 

```js
customElements.define('card-component',
  class extends HTMLElement {
    constructor() {
      super(); // Always call super first in constructor
      const cardComponent = document.getElementById('card-component-template').content;
      const shadowRoot = this.attachShadow({
        mode: 'open'
      });
      shadowRoot.appendChild(cardComponent.cloneNode(true));
    }
  });
  ```

example pen: https://codepen.io/estelle/pen/YzjNVjM

bsmth commented:
Hi @KonnorRogers - thanks a lot for the contribution. There's some review comments pending for you to take a look at before this PR can progress, do you plan to come back to this? Thanks :)

KonnorRogers commented:
@bsmth Totally forgot about this 🙈 I'll take a look and come back and fix it.

bsmth commented:
> @bsmth Totally forgot about this 🙈 I'll take a look and come back and fix it.

No problem, I'm just checking in on PRs awaiting response to see if we can unblock them, thanks! :)

KonnorRogers committed:
Update index.md

KonnorRogers committed:
Update index.md

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers commented:
Updated!

estelle committed:
Merge branch 'main' into konnorrogers/add-examples-to-exportparts

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
````diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+}
+```
+
+### Renaming / Remapping parts
````
```suggestion
### Exposing mapped parts
```

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+
+### Renaming / Remapping parts
+
+To rename an exportpartm the syntax looks something like this: `exportparts="<original-name>:<new-name>,<original-name>:<new-name>"`
```
```suggestion
To rename exported parts, we include a comma-separated list of mapped parts, with each mapped part including the original name and exported name separated by a colon (`:`): 
```

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+
+To rename an exportpartm the syntax looks something like this: `exportparts="<original-name>:<new-name>,<original-name>:<new-name>"`
+
+And here's the prior `CardWrapper` with the remapping syntax:
```
```suggestion
#### HTML

We update the prior `<card-wrapper>` custom element with the remapping syntax (omitting `body` from the exported parts list):
```

we don't explain that `CardWrapper` is the same as `<card-wrapper>`, so let's keep it as an element.

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
````diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+  </card-component>
+</template>
+```
+
````
```suggestion

#### JavaScript
```

also include the hidden JS and HTML

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
````diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+);
+```
+
+Now we can target parts on the `<card-component>` from the `<card-wrapper>` like so:
````
```suggestion
#### CSS

In targetting parts of the `<card-component>` from within the `<card-wrapper>`, we can only style the exported parts via their exposed part names:

```

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
````diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+Now we can target parts on the `<card-component>` from the `<card-wrapper>` like so:
+
+```css
+card-wrapper::part(card__header) {
+  font-weight: bold;
+}
````
```suggestion
/* selects the exported parts name */
card-wrapper::part(card__header) {
  font-weight: bold;
}
/* selects nothing: these part names were not exported */
card-wrapper::part(footer),
::part(body) {
  font-weight: bold;
}
```

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+    exportparts="
+       base:card__base, 
+       header:card__header, 
+       body:card__body, 
```
```suggestion
```
Let's continue to omit body so we can show what happens when omitted.

estelle reviewed (commented):
Thanks for all your work on this. This page was definitely missing an example, and this is a good one. It is coming along well! 

I made several suggestions to make this more understandable to people who don't know this stuff at all and to fit into how MDN writes examples

We should port some of this to the `parts` attribute page once this is complete; that page really needs help, and the intro explanation your wrote for this page will work well there.

KonnorRogers commented:
@estelle thank you so much for the thorough review and the much more approachable phrasing!

I'll work on making these changes tomorrow!

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

KonnorRogers commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -17,6 +17,122 @@ The global attribute [`part`](/en-US/docs/Web/HTML/Global_attributes#part) makes
+    }
+  </style>
+  <div class="base" part="base">
+    <div part="header"><slot name="header"></slot></div>
+    <div part="body"><slot></slot></div>
+    <div part="footer"><slot name="footer"></slot></div>
```
@estelle I did. I was using it as a "default" slot. But could see where not being explicit here could be confusing.

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

KonnorRogers committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

KonnorRogers committed:
Update index.md

KonnorRogers committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

KonnorRogers committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

KonnorRogers commented:
@estelle Sorry for taking a while, thank you so much for taking the time to make this content read so much nicer!

I applied your suggestions and removed the `` ```css`` at the end of that code block.

Let me know if there's anything else I can do to help push this across the finish line!

KonnorRogers committed:
Update index.md

KonnorRogers committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

estelle committed:
Merge branch 'main' into konnorrogers/add-examples-to-exportparts

estelle committed:
Merge branch 'main' into konnorrogers/add-examples-to-exportparts

estelle reviewed (commented):
i am going to add a commit to enable the other two examples to show.

estelle committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

estelle committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

estelle committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

estelle committed:
Example results added

estelle committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

estelle committed:
remove dash before "hidden"

estelle committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

estelle committed:
Merge branch 'main' into konnorrogers/add-examples-to-exportparts

estelle commented:
I directly committed some code to show the results. I figured showing you how to do it would be faster than explaining it. , The results are showing, but they aren't what we were expecting. I am going to try troubleshooting in CodePen.

estelle committed:
end code

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -30,6 +30,261 @@ When exporting a `part`, you have the option to assign a different name to the p
+  font-style: italic;
+}
+
+::part(header, footer) {
```
```suggestion
::part(header),
::part(footer) {
```

estelle committed:
part selector on 2 lines

estelle committed:
make ex 2 and 3 show wrapper

estelle committed:
nested slots

estelle commented:
> Other than the slots not being populated on a nested component, those examples are working. Now i need to trouble shoot that.

I figured it out with the help of @OnkarRuikar. Thank you, Onkar!!!!

see https://codepen.io/estelle/pen/qBvKyEv

estelle committed:
shadowroot styles

less line spacing.
removed grid as not needed

estelle commented:
@KonnorRogers I updated the examples so that they are displayed and added the markup needed to nest the components. Are you OK with these changes?

estelle committed:
Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

estelle commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
+
+#### Results
+
+{{ EmbedLiveSample('Nested_component', '100%', '160') }}
```
```suggestion
{{ EmbedLiveSample('Nested_component', '100%', '400') }}
```

estelle committed:
Update files/en-us/web/html/global_attributes/exportparts/index.md

OnkarRuikar commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
+  <style>
+    p {
+      margin: 0.5em 0;
+    }
```
```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```

OnkarRuikar commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
+    p,
+    h2 {
+      margin: 0.5em 0;
+    }
```
```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```

estelle replied:
i added this to the component from the main DOM and added :host display block here instead; showing a few more examples of how to bypass the DOM / shadowDOM barrier

OnkarRuikar commented on files/en-us/web/html/global_attributes/exportparts/index.md:
```diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
+  <style>
+    p {
+      margin: 0.5em 0;
+    }
```
```suggestion
    }

    .base {
      width: fit-content;
      border: 2px dashed blue;
    }
```

OnkarRuikar commented on files/en-us/web/html/global_attributes/exportparts/index.md:
````diff
@@ -30,6 +30,284 @@ When exporting a `part`, you have the option to assign a different name to the p
+
+```html
+<template id="card-wrapper">
+  <card-component exportparts="base, header, body">
+    <span slot="header_slot"><slot name="H"></slot></span>
+    <span slot="body_slot"><slot name="B"></slot></span>
+    <span slot="footer_slot"><slot name="F"></slot></span>
+  </card-component>
````
```suggestion
  <style>
    .wrapper {
      padding: 1rem;
      border: 2px dashed red;
    }
  </style>
  <div class="wrapper">
    <card-component exportparts="base, header, body">
      <span slot="header_slot"><slot name="H"></slot></span>
      <span slot="body_slot"><slot name="B"></slot></span>
      <span slot="footer_slot"><slot name="F"></slot></span>
    </card-component>
  </div>
```

OnkarRuikar reviewed (commented):
Would be great if we add a border to the card and wrapper. It would look like this:
![_1](https://github.com/mdn/content/assets/87750369/c1f4b350-c437-4ef5-abc6-6a7ef43bbfff)

estelle committed:
remove css from components

The css was for the h2, which is NOT in the shadowDOM.

estelle committed:
add :host pseudo class

displays block only when rendered

estelle committed:
host display block

KonnorRogers commented:
> I directly committed some code to show the results. I hope that is OK.
> 
> @KonnorRogers I updated the examples so that they are displayed and added the markup needed to nest the components. Are you OK with these changes?

Totally fine with me!! No issues on my end! I appreciate you taking the time to make this happen!

estelle reviewed (approved):
Yay! Thank you for making this happen. I learned so much working with you on this. 

Also, congratulations on your first merged MDN content PR. 
Welcome to the team!

KonnorRogers commented:
❤️ Thank you @estelle so much for taking the time to flesh out my half baked PR and turn it into something much more meaningful 🙈  and thank you to everyone else who helped!
